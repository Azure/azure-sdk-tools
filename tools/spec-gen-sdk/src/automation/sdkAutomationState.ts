/**
 * The generation status strings for an SDK repository.
 */
// tslint:disable-next-line: variable-name
const SDKAutomationStateStrings = {
  /**
   * The generation process has not yet begun.
   */
  pending: `Pending`,
  /**
   * The generation process is in-progress.
   */
  inProgress: `In-Progress`,
  /**
   * The generation process has failed.
   */
  failed: `Failed`,
  /**
   * The generation process has succeeded.
   */
  succeeded: `Succeeded`,
  /**
   * The generation process has warnings.
   */
  warning: `Warning`,
  /**
   * The generation process exited due to missing language configuration in tspconfig.yaml or reamd.md .
   */
  notEnabled: `NotEnabled`
};

/**
 * Get the string representation of the provided state.
 * @param state The state to get the string for.
 */
export function getSDKAutomationStateString(state: SDKAutomationState): string {
  return SDKAutomationStateStrings[state];
}

/**
 * The names of the images that associate with the different SDK Automation states.
 */
// tslint:disable-next-line: variable-name
const SDKAutomationStateImageNames = {
  /**
   * The generation process has not yet begun.
   */
  pending: `pending.gif`,
  /**
   * The generation process is in-progress.
   */
  inProgress: `inProgress.gif`,
  /**
   * The generation process has failed.
   */
  failed: `failed.gif`,
  /**
   * The generation process has succeeded.
   */
  succeeded: `succeeded.gif`,
  /**
   * The generation process has warnings.
   */
  warning: `warning.gif`
};

/**
 * Get the image name for the provided state.
 * @param state The state to get the image name for.
 */
export function getSDKAutomationStateImageName(state: SDKAutomationState): string {
  return SDKAutomationStateImageNames[state];
}

/**
 * The generation status for an SDK repository.
 */
export type SDKAutomationState = keyof typeof SDKAutomationStateStrings;

/**
 * Get the possible values for an SDKAutomationState.
 */
export function getSDKAutomationStates(): SDKAutomationState[] {
  return Object.keys(SDKAutomationStateStrings) as SDKAutomationState[];
}
