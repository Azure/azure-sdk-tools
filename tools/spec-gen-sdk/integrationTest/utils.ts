import * as fs from 'fs-extra';
import _rimraf from 'rimraf';
import { Octokit } from '@octokit/rest';
import { sdkAutomationCliConfig } from '../src/cli/config';
import { getAuthenticatedOctokit } from '../src/utils/githubUtils';
import { deleteFolder } from '@ts-common/azure-js-dev-tools';
import * as winston from 'winston';
import path from 'path';
import simpleGit, { ResetMode, SimpleGit, SimpleGitOptions } from 'simple-git';
import { getCurrentBranch, gitAddAll, gitGetCommitter, gitGetDiffFileList, gitSetRemoteWithAuth } from '../src/utils/gitUtils';
import { sdkAutoLogLevels } from '../src/automation/logging';
import { simpleGitOptions } from '../src/automation/workflow';

const fixturePrefix = 'integrationTest/fixtures';
export const workPrefix = sdkAutomationCliConfig.workingFolder;
export const repoOwner = sdkAutomationCliConfig.specRepo.split('/')[0];

let runId = sdkAutomationCliConfig.testRunId;
if (!runId) {
  runId = Math.random().toString(36).substring(2, 8);
}
export const getRunIdPrefix = () => `test-${runId}`;
const logger = winston.createLogger({
  levels: sdkAutoLogLevels.levels,
  transports: [new winston.transports.Console({ level: 'error' })]
});
const [githubClient, getGithubAccessToken] = getAuthenticatedOctokit(
  {
    token: sdkAutomationCliConfig.githubToken,
    id: sdkAutomationCliConfig.githubApp.id,
    privateKey: sdkAutomationCliConfig.githubApp.privateKey
  },
  logger
);

export const getTestGithubClient = async (): Promise<Octokit> => {
  return githubClient;
};

export const getTestGitClient = async (repoName: string) => {
  const repoFolder = path.join(workPrefix, repoName);
  fs.mkdirpSync(repoFolder);

  const repo = simpleGit({ ...simpleGitOptions, baseDir: repoFolder });
  await repo.init(false);
  return repo;
};

export const cleanTestRepos = async (repos: string[]) => {
  const github = await getTestGithubClient();

  for (const repoName of repos) {
    console.log(`Cleaning up ${repoOwner}/${repoName}`);
    await deleteFolder(`${workPrefix}/${repoName}`);
    try {
      await github.repos.delete({ owner: repoOwner, repo: repoName });
    } catch (e) {
      console.log(`Failed to delete ${repoOwner}/${repoName}: ${e.message} ${e.stack}`);
    }
  }
};

type ReplaceFileRecord = {
  [filePath: string]: {
    search: string;
    replace: string;
  }[];
};
export const applyReplaceFileRecord = (workPath: string, toReplace?: ReplaceFileRecord) => {
  if (toReplace !== undefined) {
    for (const [filePath, replaceArr] of Object.entries(toReplace)) {
      let content = fs.readFileSync(`${workPath}/${filePath}`).toString();
      for (const { search, replace } of replaceArr) {
        content = content.replace(search, replace);
      }
      fs.writeFileSync(`${workPath}/${filePath}`, content);
    }
  }
};

export const initializeGithubRepoFromLocalFixture = async (
  fixtureName: string,
  repoName: string,
  toReplace?: ReplaceFileRecord,
  baseBranch: string = 'master'
) => {
  console.log(`Initializing https://github.com/${repoOwner}/${repoName}`);
  const fixturePath = path.join(fixturePrefix, fixtureName);
  const workPath = path.join(workPrefix, repoName);
  const github = await getTestGithubClient();

  // Copy file
  await fs.remove(workPath);
  await fs.mkdirp(workPath);
  await fs.copy(fixturePath, workPath);
  applyReplaceFileRecord(workPath, toReplace);
  const repo: SimpleGit = await getTestGitClient(repoName);

  // Create repo if not exist
  try {
    await github.repos.delete({ owner: repoOwner, repo: repoName });
  } catch (e) {
    // Repo not found. Pass
  }
  await github.repos.createInOrg({ name: repoName, org: repoOwner });
  try {
    await github.repos.get({ owner: repoOwner, repo: repoName });
  } catch (e) {}

  // Checkout branch
  await gitGetCommitter(repo);
  await repo.add([]).commit(`Init repo`)
  const commit = await repo.revparse('HEAD');
  const currentBranch = await getCurrentBranch(repo);
  if (baseBranch !== currentBranch) {
    await repo.raw(['branch', '--copy', commit, baseBranch, '--force']);
    await repo.checkoutLocalBranch(baseBranch);
  }

  // Add and commit
  await gitAddAll(repo);
  const diff = await repo.diff(['--name-status', 'HEAD']);
  let fileList = await gitGetDiffFileList(diff);
  await repo.add(fileList).commit(`Init repo with fixture ${fixtureName}`)

  // Push
  await gitSetRemoteWithAuth({ logger, getGithubAccessToken }, repo, 'origin', { owner: repoOwner, name: repoName });
  await repo.push([
    'origin',
    `+refs/heads/${baseBranch}:refs/heads/${baseBranch}`
  ]);
  console.log(`Initialized https://github.com/${repoOwner}/${repoName}`);
};

export const createPRWithPatch = async (
  repoName: string,
  patchName: string,
  branchName: string,
  options: {
    toReplace?: ReplaceFileRecord;
    baseBranch?: string;
    fetchBase?: boolean;
  } = {}
) => {
  const baseBranch = options.baseBranch ?? 'master';

  console.log(`Creating PR in '${repoOwner}/${repoName}' with patch '${patchName}' on branch '${branchName}'`);
  const workPath = `${workPrefix}/${repoName}`;
  const patchPath = `${fixturePrefix}/${patchName}`;
  const github = await getTestGithubClient();
  const repo = await getTestGitClient(repoName);

  await gitAddAll(repo);
  const header = await repo.revparse('HEAD')
  await repo.reset(ResetMode.HARD, [header]);
  await repo.checkout(header, ['--force']);

  if (options.fetchBase) {
    await repo.fetch([
       'origin',
       '--update-head-ok',
       '--force',
       `+refs/heads/${baseBranch}:refs/remotes/origin/${baseBranch}`
     ]);
    await repo.mergeFromTo(`origin/${baseBranch}`, baseBranch);
  }

  await repo.checkout(branchName, ['--force']);
  const baseCommit = await repo.revparse('HEAD');
  await repo.raw(['branch', '--copy', baseCommit, baseBranch, '--force']);
  await repo.checkout(branchName, ['--force']);

  await fs.copy(patchPath, workPath);
  applyReplaceFileRecord(workPath, options.toReplace);

  await gitAddAll(repo);
  const diff = await repo.diff(['--name-status', 'HEAD']);
  let fileList = await gitGetDiffFileList(diff);
  await gitGetCommitter(repo);
  await repo.add(fileList).commit(`Apply patch ${patchName}`)

  // Push
  await gitSetRemoteWithAuth({ logger, getGithubAccessToken }, repo, 'origin', { owner: repoOwner, name: repoName });
  await repo.push([
    'origin',
    `+refs/heads/${branchName}:refs/heads/${branchName}`
  ]);

  let pr = await github.pulls.create({
    owner: repoOwner,
    repo: repoName,
    head: branchName,
    base: baseBranch,
    title: branchName
  });
  console.log(pr.data.html_url);

  // Wait for mergable data
  while (pr.data.mergeable === null) {
    pr = await github.pulls.get({ owner: repoOwner, repo: repoName, pull_number: pr.data.number });
  }

  return pr.data.number;
};

export const getPRListInfo = async (repoName: string) => {
  const github = await getTestGithubClient();
  // Wait for 5 seconds to make sure we could get the latest pull request list
  await new Promise((resolve) => setTimeout(resolve, 5000));
  const { data: sdkPRs } = await github.pulls.list({
    owner: repoOwner,
    repo: repoName,
    state: 'all'
  });
  const result = sdkPRs.map((pr) => ({
    number: pr.number,
    state: pr.state,
    title: pr.title,
    base: pr.base.ref,
    head: pr.head.ref
  }));
  result.sort((a, b) => a.number - b.number);
  return result;
};

export const getLoggerTextForAssert = async (
  _logger: { allLogs: string[] },
  specRepoName: string,
  specPRNumber: number,
  sdkRepoName: string
) => {
  let text = _logger.allLogs.join('\n');
  text = text.replace(new RegExp(fs.realpathSync('.'), 'g'), '.');
  text = text.replace(new RegExp(repoOwner, 'g'), '<org>');
  text = text.replace(new RegExp(specRepoName, 'g'), '<spec_repo>');
  text = text.replace(new RegExp(sdkRepoName, 'g'), '<sdk_repo>');

  text = text.replace(/ [a-z0-9]{40}/g, ` <commit_sha>`);

  const version = require('pkginfo').read(module).package.version;
  text = text.replace(new RegExp(version.replace('.', '\\.'), 'g'), '<version>');

  text = text.replace(new RegExp(path.resolve(workPrefix), 'g'), '<work_dir>');

  return text;
};

export const launchTestSDKAutomation = async (automationPromise: Promise<void>, info: string) => {
  console.log(`Launching SDK Automation for ${info}`);
  try {
    await automationPromise;
  } catch (e) {
    console.log(e);
    // Need log for diagnostic so do not fail on exception
  }
};
