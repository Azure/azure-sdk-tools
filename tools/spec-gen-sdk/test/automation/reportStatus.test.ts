vi.mock('node:fs', () => ({
  existsSync: vi.fn(),
  readFileSync: vi.fn(() => Buffer.from('{{mock handlebars template}}')),
  rmSync: vi.fn(),
  writeFileSync: vi.fn(),
}));

vi.mock('node:path', () => ({
  ...vi.importActual('node:path'),
  join: vi.fn((...args) => args.join('/')),
}));

vi.mock('../../src/utils/runScript');
vi.mock('../../src/utils/utils');
vi.mock('../../src/automation/logging');
vi.mock('../../src/utils/fsUtils');
vi.mock('../../src/utils/messageUtils');
vi.mock('../../src/utils/workflowUtils');
vi.mock('../../src/utils/reportStatusUtils');
vi.mock('marked', () => ({
  marked: vi.fn(() => '<p>mocked markdown</p>'),
}));

import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import * as fs from 'node:fs';
import { generateReport, saveVsoLog, saveFilteredLog, generateHtmlFromFilteredLog } from '../../src/automation/reportStatus';
import { WorkflowContext, FailureType } from '../../src/types/Workflow';
import { SDKAutomationState } from '../../src/automation/sdkAutomationState';
import { PackageData } from '../../src/types/PackageData';

// Import mocked modules
import { setSdkAutoStatus } from '../../src/utils/runScript';
import { extractPathFromSpecConfig, mapToObject } from '../../src/utils/utils';
import { vsoAddAttachment, vsoLogError, vsoLogWarning } from '../../src/automation/logging';
import { deleteTmpJsonFile, writeTmpJsonFile } from '../../src/utils/fsUtils';
import { toolError, toolWarning } from '../../src/utils/messageUtils';
import { setFailureType } from '../../src/utils/workflowUtils';
import { commentDetailView, renderHandlebarTemplate } from '../../src/utils/reportStatusUtils';
import { marked } from 'marked';

describe('reportStatus', () => {
  let mockWorkflowContext: WorkflowContext;
  let mockLogger: any;
  let mockPackageData: PackageData;

  beforeEach(() => {
    vi.clearAllMocks();
    
    const fixedDate = new Date('2022-01-01T00:00:00Z');
    vi.spyOn(global, 'Date').mockImplementation(() => fixedDate);

    mockLogger = {
      log: vi.fn(),
      info: vi.fn(),
      warn: vi.fn(),
      error: vi.fn(),
    };

    mockPackageData = {
      name: 'test-package',
      serviceName: 'test-service',
      relativeFolderPath: 'packages/test-package',
      extraRelativeFolderPaths: [],
      status: 'succeeded' as SDKAutomationState,
      messages: [],
      hasBreakingChange: false,
      isPrivatePackage: false,
      artifactPaths: ['artifact1.zip', 'artifact2.zip'],
      readmeMd: ['README.md'],
      typespecProject: undefined,
      version: '1.0.0',
      apiViewArtifactPath: 'apiview.json',
      language: 'JavaScript',
      isBetaMgmtSdk: false,
      isDataPlane: false,
      presentSuppressionLines: [],
      absentSuppressionLines: [],
      installationInstructions: 'npm install test-package',
      liteInstallationInstruction: undefined,
      // Required fields for SDKRepositoryPackageData
      changedFilePaths: [],
      generationBranch: 'generation-branch',
      generationRepository: 'test-repo',
      generationRepositoryUrl: 'https://github.com/test/repo',
      integrationBranch: 'integration-branch',
      integrationRepository: 'test-integration-repo',
    } as unknown as PackageData;

    mockWorkflowContext = {
      config: {
        specRepo: { owner: 'Azure', name: 'azure-rest-api-specs' },
        sdkName: 'azure-sdk-for-js',
        branchPrefix: 'sdkAuto',
        localSpecRepoPath: '/spec/repo',
        localSdkRepoPath: '/sdk/repo',
        tspConfigPath: '/spec/tspconfig.yaml',
        readmePath: '/spec/README.md',
        pullNumber: '123',
        apiVersion: '2023-01-01',
        runMode: 'test',
        sdkReleaseType: 'beta',
        specCommitSha: 'abc123',
        specRepoHttpsUrl: 'https://github.com/Azure/azure-rest-api-specs',
        workingFolder: '/tmp/working',
        headRepoHttpsUrl: 'https://github.com/user/azure-rest-api-specs',
        headBranch: 'feature-branch',
        runEnv: 'azureDevOps' as const,
        version: '1.0.0',
        skipSdkGenFromOpenapi: 'false',
      },
      logger: mockLogger,
      fullLogFileName: '/tmp/full.log',
      filteredLogFileName: '/tmp/filtered.log',
      htmlLogFileName: '/tmp/report.html',
      vsoLogFileName: '/tmp/vso.log',
      specRepoConfig: {
        sdkRepositoryMappings: {},
        overrides: {},
        typespecEmitterToSdkRepositoryMapping: {},
      },
      sdkRepoConfig: {
        configFilePath: 'swagger_to_sdk_config.json',
        mainRepository: { owner: 'test', name: 'test-repo' },
        mainBranch: 'main',
        integrationRepository: { owner: 'test', name: 'test-integration-repo' },
        secondaryRepository: { owner: 'test', name: 'test-secondary-repo' },
        secondaryBranch: 'secondary',
        integrationBranchPrefix: 'integration-',
      },
      swaggerToSdkConfig: {
        packageOptions: {
          breakingChangesLabel: 'breaking-change',
        },
      },
      isPrivateSpecRepo: false,
      stagedArtifactsFolder: '/tmp/artifacts',
      sdkArtifactFolder: '/tmp/sdk-artifacts',
      isSdkConfigDuplicated: false,
      specConfigPath: 'specification/test/resource-manager',
      pendingPackages: [],
      handledPackages: [mockPackageData],
      status: 'succeeded' as SDKAutomationState,
      failureType: undefined,
      messages: [],
      messageCaptureTransport: {} as any,
      scriptEnvs: {},
      tmpFolder: '/tmp',
      vsoLogs: new Map(),
    } as unknown as WorkflowContext;

    // Mock external dependencies
    vi.mocked(extractPathFromSpecConfig).mockReturnValue('test-spec-config');
    vi.mocked(mapToObject).mockReturnValue({});
    vi.mocked(toolError).mockImplementation((msg) => `ERROR: ${msg}`);
    vi.mocked(toolWarning).mockImplementation((msg) => `WARNING: ${msg}`);
    vi.mocked(renderHandlebarTemplate).mockReturnValue('mocked comment body');
    // commentDetailView will be the mocked compiled function from the module
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  describe('generateReport', () => {
    it('should generate report successfully with valid context', () => {
      vi.mocked(fs.existsSync).mockReturnValue(false);

      const writeTmpJsonFileMock = vi.fn();
      vi.mocked(writeTmpJsonFile).mockImplementation(writeTmpJsonFileMock);
      
      generateReport(mockWorkflowContext);

      expect(mockLogger.log).toHaveBeenCalledWith('section', 'Generate report');
      expect(mockLogger.log).toHaveBeenCalledWith('endsection', 'Generate report');
      expect(setSdkAutoStatus).toHaveBeenCalledWith(mockWorkflowContext, 'succeeded');
      expect(deleteTmpJsonFile).toHaveBeenCalledWith(mockWorkflowContext, 'execution-report.json');
      expect(writeTmpJsonFileMock.mock.calls[0][2]).toMatchSnapshot();
    });

    it('should handle breaking changes correctly', () => {
      const packageWithBreakingChange = {
        ...mockPackageData,
        hasBreakingChange: true,
        isBetaMgmtSdk: false,
        isDataPlane: false,
        presentSuppressionLines: [],
        absentSuppressionLines: [],
      };
      
      mockWorkflowContext.handledPackages = [packageWithBreakingChange];

      generateReport(mockWorkflowContext);

      expect(writeTmpJsonFile).toHaveBeenCalledWith(
        mockWorkflowContext,
        'execution-report.json',
        expect.objectContaining({
          packages: expect.arrayContaining([
            expect.objectContaining({
              shouldLabelBreakingChange: true,
              areBreakingChangeSuppressed: false,
            }),
          ]),
        })
      );
    });

    it('should handle suppressed breaking changes correctly', () => {
      const packageWithSuppressedBreakingChange = {
        ...mockPackageData,
        hasBreakingChange: true,
        isBetaMgmtSdk: false,
        isDataPlane: false,
        presentSuppressionLines: ['suppression1', 'suppression2'],
        absentSuppressionLines: [],
      };
      
      mockWorkflowContext.handledPackages = [packageWithSuppressedBreakingChange];

      generateReport(mockWorkflowContext);

      expect(writeTmpJsonFile).toHaveBeenCalledWith(
        mockWorkflowContext,
        'execution-report.json',
        expect.objectContaining({
          packages: expect.arrayContaining([
            expect.objectContaining({
              shouldLabelBreakingChange: false,
              areBreakingChangeSuppressed: true,
            }),
          ]),
        })
      );
    });

    it('should write markdown file when pullNumber is provided', () => {
      vi.mocked(fs.existsSync).mockReturnValue(false);

      const writeFileSyncMock = vi.fn();
      vi.mocked(fs.writeFileSync).mockImplementation(writeFileSyncMock);

      generateReport(mockWorkflowContext);
      
      expect(writeFileSyncMock.mock.calls[0][0]).toBe('/tmp/working/out/logs/test-spec-config-package-report.md');
      expect(writeFileSyncMock.mock.calls[0][1]).toMatchSnapshot();
      expect(vsoAddAttachment).toHaveBeenCalledWith(
        'Generation Summary for specification-test-resource-manager',
        '/tmp/working/out/logs/test-spec-config-package-report.md'
      );
    });
    
    it('should remove existing markdown file before writing new one', () => {
      vi.mocked(fs.existsSync).mockReturnValue(true);

      generateReport(mockWorkflowContext);

      expect(fs.rmSync).toHaveBeenCalledWith('/tmp/working/out/logs/test-spec-config-package-report.md');
      expect(fs.writeFileSync).toHaveBeenCalled();
    });

    it('should handle markdown write errors', () => {
      const writeError = new Error('Write failed');
      vi.mocked(fs.writeFileSync).mockImplementation(() => {
        throw writeError;
      });

      generateReport(mockWorkflowContext);

      expect(toolError).toHaveBeenCalledWith(expect.stringContaining('Write failed'));
      expect(mockLogger.error).toHaveBeenCalledWith(expect.stringContaining('Write failed'));
      expect(vsoLogError).toHaveBeenCalled();
    });

    it('should skip markdown generation when no pullNumber', () => {
      mockWorkflowContext.config.pullNumber = undefined;

      generateReport(mockWorkflowContext);

      expect(fs.writeFileSync).not.toHaveBeenCalled();
      expect(vsoAddAttachment).not.toHaveBeenCalled();
    });

    it('should handle local run environment (no vsoLogPath)', () => {
      mockWorkflowContext.config.runEnv = 'local';

      generateReport(mockWorkflowContext);

      expect(writeTmpJsonFile).toHaveBeenCalledWith(
        mockWorkflowContext,
        'execution-report.json',
        expect.not.objectContaining({
          vsoLogPath: expect.anything(),
        })
      );
    });
  });

  describe('saveVsoLog', () => {
    it('should save VSO log successfully', () => {
      const mockVsoLogs = new Map([
        ['task1', { errors: ['error1'], warnings: ['warning1'] }],
        ['task2', { errors: ['error2'] }],
      ]);

      mockWorkflowContext.vsoLogs = mockVsoLogs;
      vi.mocked(mapToObject).mockReturnValue({ task1: { errors: ['error1'], warnings: ['warning1'] } });

      saveVsoLog(mockWorkflowContext);

      expect(mockLogger.log).toHaveBeenCalledWith('section', expect.stringContaining(mockWorkflowContext.vsoLogFileName));
      expect(mockLogger.log).toHaveBeenCalledWith('endsection', expect.stringContaining(mockWorkflowContext.vsoLogFileName));
      expect(mapToObject).toHaveBeenCalledWith(mockVsoLogs);
      expect(fs.writeFileSync).toHaveBeenCalledWith(
        mockWorkflowContext.vsoLogFileName,
        JSON.stringify({ task1: { errors: ['error1'], warnings: ['warning1'] } }, null, 2)
      );
    });

    it('should handle write errors', () => {
      const writeError = new Error('Write failed');
      vi.mocked(fs.writeFileSync).mockImplementation(() => {
        throw writeError;
      });

      saveVsoLog(mockWorkflowContext);

      expect(toolError).toHaveBeenCalledWith(expect.stringContaining('Write failed'));
      expect(mockLogger.error).toHaveBeenCalledWith(expect.stringContaining('Write failed'));
    });
  });

  describe('saveFilteredLog', () => {
    it('should save filtered log successfully', () => {
      saveFilteredLog(mockWorkflowContext);

      expect(mockLogger.log).toHaveBeenCalledWith('section', 'Save filtered log');
      expect(mockLogger.log).toHaveBeenCalledWith('endsection', 'Save filtered log status');
      expect(renderHandlebarTemplate).toHaveBeenCalledWith(
        commentDetailView,
        mockWorkflowContext,
        expect.objectContaining({
          hasBreakingChange: false,
          showLiteInstallInstruction: false,
        })
      );
      const writeFileSyncMock = vi.mocked(fs.writeFileSync)
      expect(writeFileSyncMock.mock.calls[0][0]).toBe('/tmp/filtered.log');
      expect(writeFileSyncMock.mock.calls[0][1]).toMatchSnapshot();
    });

    it('should handle pending packages', () => {
      const pendingPackage = { ...mockPackageData, name: 'pending-package' };
      mockWorkflowContext.pendingPackages = [pendingPackage];

      saveFilteredLog(mockWorkflowContext);

      expect(setSdkAutoStatus).toHaveBeenCalledWith(mockWorkflowContext, 'failed');
      expect(setFailureType).toHaveBeenCalledWith(mockWorkflowContext, FailureType.SpecGenSdkFailed);
      expect(toolError).toHaveBeenCalled();
      expect(mockLogger.error).toHaveBeenCalledTimes(2);
      expect(mockWorkflowContext.handledPackages).toContain(pendingPackage);
    });

    it('should detect breaking changes across packages', () => {
      const packageWithBreakingChange = {
        ...mockPackageData,
        hasBreakingChange: true,
        liteInstallationInstruction: 'lite install command',
      };
      mockWorkflowContext.handledPackages = [packageWithBreakingChange];

      saveFilteredLog(mockWorkflowContext);

      expect(renderHandlebarTemplate).toHaveBeenCalledWith(
        commentDetailView,
        mockWorkflowContext,
        expect.objectContaining({
          hasBreakingChange: true,
          showLiteInstallInstruction: true,
        })
      );
    });

    it('should map different status levels correctly', () => {
      const testCases = [
        { status: 'pending' as SDKAutomationState, expectedLevel: 'Error' },
        { status: 'inProgress' as SDKAutomationState, expectedLevel: 'Error' },
        { status: 'failed' as SDKAutomationState, expectedLevel: 'Error' },
        { status: 'warning' as SDKAutomationState, expectedLevel: 'Warning' },
        { status: 'succeeded' as SDKAutomationState, expectedLevel: 'Info' },
        { status: 'notEnabled' as SDKAutomationState, expectedLevel: 'Warning' },
      ];

      testCases.forEach(({ status, expectedLevel }) => {
        vi.clearAllMocks();
        mockWorkflowContext.status = status;

        saveFilteredLog(mockWorkflowContext);

        
      const writeFileSyncMock = vi.mocked(fs.writeFileSync)
      expect(writeFileSyncMock.mock.calls[0][0]).toBe('/tmp/filtered.log');
      expect(writeFileSyncMock.mock.calls[0][1]).toMatchSnapshot();
      });
    });

    it('should handle write errors', () => {
      const writeError = new Error('Write failed');
      vi.mocked(fs.writeFileSync).mockImplementation(() => {
        throw writeError;
      });

      saveFilteredLog(mockWorkflowContext);

      expect(toolError).toHaveBeenCalledWith(expect.stringContaining('Write failed'));
      expect(mockLogger.error).toHaveBeenCalledWith(expect.stringContaining('Write failed'));
      expect(vsoLogError).toHaveBeenCalled();
    });
  });

  describe('generateHtmlFromFilteredLog', () => {
    beforeEach(() => {
      const mockMessageRecord = {
        type: 'Markdown',
        mode: 'replace',
        level: 'Info',
        message: '> [!NOTE]\n> This is a note block\n\n<ul><li>List item 1</li><li>List item 2</li></ul>',
        time: new Date(),
      };
      vi.mocked(fs.readFileSync).mockReturnValue(Buffer.from(JSON.stringify(mockMessageRecord)));
      vi.mocked(marked).mockReturnValue('<p>processed markdown</p>');
    });

    it('should generate HTML from filtered log successfully', () => {
      generateHtmlFromFilteredLog(mockWorkflowContext);

      expect(fs.readFileSync).toHaveBeenCalledWith('/tmp/filtered.log');
      const writeFileSyncMock = vi.mocked(fs.writeFileSync);
      expect(writeFileSyncMock.mock.calls[0][0]).toBe('/tmp/report.html');
      expect(writeFileSyncMock.mock.calls[0][1]).toMatchSnapshot();
      expect(writeFileSyncMock.mock.calls[0][2]).toBe('utf-8');
    });

    it('should handle markdown without note blocks', () => {
      const mockMessageRecord = {
        type: 'Markdown',
        mode: 'replace',
        level: 'Info',
        message: '# Simple Markdown\n\nThis is a simple paragraph.',
        time: new Date(),
      };
      vi.mocked(fs.readFileSync).mockReturnValue(Buffer.from(JSON.stringify(mockMessageRecord)));

      generateHtmlFromFilteredLog(mockWorkflowContext);

      const markedMock = vi.mocked(marked).mock.calls[0][0];
      expect(markedMock).toMatchSnapshot();
    });

    it('should handle read errors', () => {
      const readError = new Error('Read failed');
      vi.mocked(fs.readFileSync).mockImplementation(() => {
        throw readError;
      });

      generateHtmlFromFilteredLog(mockWorkflowContext);

      expect(toolError).toHaveBeenCalledWith(expect.stringContaining('Read failed'));
      expect(mockLogger.error).toHaveBeenCalledWith(expect.stringContaining("Read failed"));
      expect(vsoLogError).toHaveBeenCalled();
    });

    it('should handle JSON parse errors', () => {
      vi.mocked(fs.readFileSync).mockReturnValue(Buffer.from('invalid json'));

      generateHtmlFromFilteredLog(mockWorkflowContext);

      expect(toolError).toHaveBeenCalled();
      expect(mockLogger.error).toHaveBeenCalled();
      expect(vsoLogError).toHaveBeenCalled();
    });

    it('should handle write errors', () => {
      const writeError = new Error('Write failed');
      vi.mocked(fs.writeFileSync).mockImplementation(() => {
        throw writeError;
      });

      generateHtmlFromFilteredLog(mockWorkflowContext);

      expect(toolError).toHaveBeenCalledWith(expect.stringContaining('Write failed'));
      expect(mockLogger.error).toHaveBeenCalledWith(expect.stringContaining('Write failed'));
      expect(vsoLogError).toHaveBeenCalled();
    });

    it('should generate correct page title for different SDK names', () => {
      mockWorkflowContext.config.sdkName = 'azure-sdk-for-python';

      generateHtmlFromFilteredLog(mockWorkflowContext);

      const htmlContent = vi.mocked(fs.writeFileSync).mock.calls[0][1] as string;
      expect(htmlContent).toContain('<title>spec-gen-sdk-python result</title>');
    });

  });
});
