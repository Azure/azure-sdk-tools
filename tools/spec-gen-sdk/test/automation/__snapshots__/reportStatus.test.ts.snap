// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`reportStatus > generateHtmlFromFilteredLog > should parse note blocks correctly 1`] = `
"
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>spec-gen-sdk-js result</title>
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/github-markdown-css/5.2.0/github-markdown.min.css">
    <style>
        body {
            font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji";
            line-height: 1.6;
            padding: 40px;
        }
        .markdown-body {
            box-sizing: border-box;
            min-width: 200px;
            max-width: 980px;
            margin: 0 auto;
        }
        table {
            border-collapse: collapse;
            border-spacing: 0;
        }
        a {
            text-decoration: underline!important;
            text-underline-offset: .2rem!important;
        }
         /* GitHub Special prompt block */
        .markdown-alert.markdown-alert-note {
            border-left-color: #0969da;
        }
        .markdown-alert.markdown-alert-note .markdown-alert-title {
            color: #0969da;
        }
        .markdown-body>*:first-child {
            margin-top: 0 !important;
        }
        .markdown-alert {
           padding: 0.5rem 1rem;
           margin-bottom: 1rem;
           color: inherit;
           border-left: .25em solid #d1d9e0;
        }
        .markdown-alert .markdown-alert-title {
            display: flex;
            font-weight: 500;
            align-items: center;
            line-height: 1;
        }
        .markdown-alert>:first-child {
            margin-top: 0;
        }
        .markdown-body p, .markdown-body blockquote, .markdown-body ul, .markdown-body ol, .markdown-body dl, .markdown-body table, .markdown-body pre, .markdown-body details {
            margin-top: 0;
            margin-bottom: 1rem;
        }
        .mr-2 {
            margin-right: 0.5rem !important;
        }
        .octicon {
            display: inline-block;
            overflow: visible !important;
            vertical-align: text-bottom;
            fill: currentColor;
        }
    </style>
</head>
<body>
    <article class="markdown-body">
        
<div class="markdown-alert markdown-alert-note" dir="auto">
    <p class="markdown-alert-title" dir="auto">
    <svg class="octicon octicon-info mr-2" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.5 7.75A.75.75 0 0 1 7.25 7h1a.75.75 0 0 1 .75.75v2.75h.25a.75.75 0 0 1 0 1.5h-2a.75.75 0 0 1 0-1.5h.25v-2h-.25a.75.75 0 0 1-.75-.75ZM8 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"></path></svg>
        Note
    </p>
    <p dir="auto">This is a note block</p>
</div>
<ul><li>List item 1</li><li>List item 2</li></ul>
    </article>
</body>
</html>
"
`;
